### PR TITLE
Standalonemmcore to support Intel X64

### PR DESCRIPTION
**Pull request by xuweiintel**
# Description

<_Include a description of the change and why this change was made._>

<_For each item, place an "x" in between `[` and `]` if true. Example: `[x]` (you can also check items in GitHub UI)_>

<_Create the PR as a Draft PR if it is only created to run CI checks._>

<_Delete lines in \<\> tags before creating the PR._>

- [ ] Breaking change?
  - **Breaking change** - Will this cause a break in build or boot behavior?
  - Examples: Add a new library class or move a module to a different repo.
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests or integration tests.

## How This Was Tested

<_Describe the test(s) that were run to verify the changes._>

## Integration Instructions

<_Describe how these changes should be integrated. Use N/A if nothing is required._>

### CODE REVIEW

File: StandaloneMmPkg/Core/Dispatcher.c [hunk 1]
```diff
@@ -185,6 +185,47 @@ MmLoadImage (
...
+             EfiRuntimeServicesData,
+             sizeof (EFI_LOADED_IMAGE_PROTOCOL),
+             (VOID **)&DriverEntry->LoadedImage
+             );
+  if (EFI_ERROR (Status)) {
+    MmFreePages (DstBuffer, PageCount);
+    return Status;
+  }
+
+  ZeroMem (DriverEntry->LoadedImage, sizeof (EFI_LOADED_IMAGE_PROTOCOL));
```
Comment (niruiyu) [inline StandaloneMmPkg/Core/Dispatcher.c]: Can we avoid allocating memory for DriverEntry->LoadedImage by adding "EFI_LOADED_IMAGE_PROTOCOL LoadedImage" field to DriverEntry structure?

File: StandaloneMmPkg/Core/StandaloneMmCore.c [hunk 2]
```diff
@@ -440,33 +500,77 @@ MmEntryPoint (
...
-      // Communicate service finished, reset the pointer to CommBuffer to NULL
+      // Synchronous MMI for MM Core or request from Communicate protocol
       //
-      gMmCorePrivate->BufferSize         += OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data);
-      gMmCorePrivate->CommunicationBuffer = 0;
-      gMmCorePrivate->ReturnStatus        = (Status == EFI_SUCCESS) ? EFI_SUCCESS : EFI_NOT_FOUND;
+      if (!MmIsBufferOutsideMmValid (
+             (UINTN)mMmCommunicationBuffer->FixedCommBuffer,
+             mMmCommunicationBuffer->FixedCommBufferSize
+             ))
```
Comment (niruiyu) [inline StandaloneMmPkg/Core/StandaloneMmCore.c:539]: You already validated the FixedCommBuffer location in MmCoreEntryPoint. It's not needed to validate again.

File: StandaloneMmPkg/Core/StandaloneMmCore.c [hunk 3]
```diff
@@ -440,33 +500,77 @@ MmEntryPoint (
...
+        //
+        CommunicationInOut->ReturnStatus = EFI_INVALID_PARAMETER;
+      } else {
+        CommunicateHeader = (EFI_MM_COMMUNICATE_HEADER *)(UINTN)mMmCommunicationBuffer->FixedCommBuffer;
+        BufferSize        = OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data) + CommunicateHeader->MessageLength;
+        if (BufferSize <= mMmCommunicationBuffer->FixedCommBufferSize) {
+          //
+          // Shadow the communication buffer to internal buffer
+          //
+          ZeroMem (mInternalCommBufferCopy, mMmCommunicationBuffer->FixedCommBufferSize);
```
Comment (niruiyu) [inline StandaloneMmPkg/Core/StandaloneMmCore.c:552]: Optimize the length of ZeroMem() to only zero out the remaining portion of buffer.

File: StandaloneMmPkg/Core/StandaloneMmCore.c [hunk 4]
```diff
@@ -440,33 +500,77 @@ MmEntryPoint (
...
+
+          BufferSize = BufferSize + OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data);
+          if (BufferSize <= mMmCommunicationBuffer->FixedCommBufferSize) {
+            //
+            // Copy the data back to FixedCommBuffer
+            //
+            ZeroMem (
+              (VOID *)(UINTN)mMmCommunicationBuffer->FixedCommBuffer,
+              mMmCommunicationBuffer->FixedCommBufferSize
+              );
```
Comment (niruiyu) [inline StandaloneMmPkg/Core/StandaloneMmCore.c:576]: No need to call ZeroMem() again.
The purpose of ZeroMem() is to not pass data from one Handler to another.

File: StandaloneMmPkg/Core/Dispatcher.c [hunk 5]
```diff
@@ -185,6 +185,47 @@ MmLoadImage (
   DriverEntry->ImageBuffer     = DstBuffer;
   DriverEntry->NumberOfPage    = PageCount;
 
+  Status = MmAllocatePool (
+             EfiRuntimeServicesData,
+             sizeof (EFI_LOADED_IMAGE_PROTOCOL),
+             (VOID **)&DriverEntry->LoadedImage
+             );
+  if (EFI_ERROR (Status)) {
+    MmFreePages (DstBuffer, PageCount);
```
Comment (hongbin123) [inline StandaloneMmPkg/Core/Dispatcher.c:194]: If it is aligned to use MmFreePages for MmAllocatePool ?

File: StandaloneMmPkg/Core/StandaloneMmCore.c [hunk 6]
```diff
@@ -679,7 +679,10 @@ StandaloneMmMain (
   //
   MmramRangesHob = GetNextGuidHob (&gEfiMmPeiMmramMemoryReserveGuid, HobStart);
   if (MmramRangesHob == NULL) {
-    return EFI_UNSUPPORTED;
+    MmramRangesHob = GetFirstGuidHob (&gEfiSmmSmramMemoryGuid);
+    if (MmramRangesHob == NULL) {
+      return EFI_UNSUPPORTED;
+    }
```
Comment (niruiyu) [inline StandaloneMmPkg/Core/StandaloneMmCore.c]: gEfiSmmSmramMemoryGuid is defined in PI spec but the gEfiMmPeiMmramMemoryReserveGuid is not. So, please locate the spec-defined GUIDed HOB first.

File: StandaloneMmPkg/Core/StandaloneMmCore.c [hunk 7]
```diff
@@ -348,46 +523,69 @@ MmEntryPoint (
...
+        //
+        // Shadow the data from MM Communication Buffer to internal buffer
+        //
+        CopyMem (
+          mInternalCommBufferCopy,
+          (VOID *)(UINTN)mMmCommunicationBuffer->PhysicalStart,
+          BufferSize
+          );
+        ZeroMem (
+          mInternalCommBufferCopy,
```
Comment (jiaxinwu) [inline StandaloneMmPkg/Core/StandaloneMmCore.c:548]: should be :
        ZeroMem (
          (UINT8 *)mInternalCommBufferCopy + BufferSize,
          EFI_PAGES_TO_SIZE (mMmCommunicationBuffer->NumberOfPages) - BufferSize
          );

File: StandaloneMmPkg/Core/StandaloneMmCore.c [hunk 8]
```diff
@@ -399,11 +597,6 @@ MmEntryPoint (
   // TBD: Do not use private data structure ?
```
Comment (jiaxinwu) [inline StandaloneMmPkg/Core/StandaloneMmCore.c:597]: Remove the TBD.

File: StandaloneMmPkg/Core/StandaloneMmCore.c [hunk 9]
```diff
@@ -643,6 +643,61 @@ MmConfigurationMmNotify (
...
+        MemoryInMmram = AllocatePages (EFI_SIZE_TO_PAGES (MemoryAllocationHob->AllocDescriptor.MemoryLength));
+        if (MemoryInMmram != NULL) {
+          DEBUG ((
+            DEBUG_INFO,
+            "Migrate Memory Allocation Hob (%g) from %08x to %08p\n",
+            &MemoryAllocationHob->AllocDescriptor.Name,
+            MemoryAllocationHob->AllocDescriptor.MemoryBaseAddress,
+            MemoryInMmram
+            ));
+          MemoryAllocationHob->AllocDescriptor.MemoryBaseAddress = (EFI_PHYSICAL_ADDRESS)(UINTN)MemoryInMmram;
```
Comment (jiaxinwu) [inline StandaloneMmPkg/Core/StandaloneMmCore.c:683]: lost the mem copy from original non-smm buffer to mm buffer?

**Pull request closed without merging**